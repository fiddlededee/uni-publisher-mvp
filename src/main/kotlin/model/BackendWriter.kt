package model
interface BackendWriter {
    // pattern visitor
    fun write(openBlock: OpenBlock)
    fun write(table: Table)
    fun write(tableRowGroup: TableRowGroup)
    fun write(tr: TableRow)
    fun write(td: TableCell)
    fun write(colGroup: ColGroup)
    fun write(col: Col)
    fun write(ol: OrderedList)
    fun write(ul: UnorderedList)
    fun write(li: ListItem)
    fun write(h: Header)
    fun write(p: Paragraph)
//    fun write(pre: Pre)
//    fun write(code: Code)
    fun write(doc: Document)
    fun write(text: Text)
    fun write(span: Span)
    fun write(a: Anchor)
    fun write(dummyNode: Node)
}
