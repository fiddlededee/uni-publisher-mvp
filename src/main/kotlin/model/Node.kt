package model

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.annotation.JsonTypeInfo
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory
import com.fasterxml.jackson.dataformat.yaml.YAMLGenerator

@JsonInclude(JsonInclude.Include.NON_EMPTY)
@JsonTypeInfo(include = JsonTypeInfo.As.WRAPPER_OBJECT, use = JsonTypeInfo.Id.NAME)
abstract class Node() {

    // todo: next/previous-sibling
    // todo: ancestors(), nextSiblings(), previousSiblings()
    // todo: is firtst, is last
    var id: String? = null


    @get:JsonIgnore
    abstract val isInline: Boolean

    @JsonIgnore
    private var parent: Node? = null

    val children: ArrayList<Node> = arrayListOf()
    val roles = arrayListOf<String>()
    fun <T : Node> addChild(childNode: T): T {
        val oldParent = childNode.parent
        children.add(childNode)
        childNode.parent = this
        oldParent?.removeChild(childNode)
        return childNode
    }

    // todo : implement all builder functions

    fun table(init: Table.() -> Unit = {}): Table {
        return addChild(Table().apply(init))
    }

    fun tableRowGroup(type: TRG, init: TableRowGroup.() -> Unit = {}): TableRowGroup {
        return addChild(TableRowGroup(type).apply(init))
    }

    fun tr(init: TableRow.() -> Unit = {}): TableRow {
        return addChild(TableRow().apply(init))
    }

    fun td(init: TableCell.() -> Unit = {}): TableCell {
        return addChild(TableCell().apply(init))
    }

    fun h(level: Int, init: Header.() -> Unit = {}): Header {
        return addChild(Header(level).apply(init))
    }


    fun li(init: ListItem.() -> Unit = {}): ListItem {
        return addChild(ListItem().apply(init))
    }

    fun p(init: Paragraph.() -> Unit = {}): Paragraph {
        return addChild(Paragraph().apply(init))
    }

    fun span(init: Span.() -> Unit = {}): Span {
        return addChild(Span().apply(init))
    }

    fun a(href: String, init: Anchor.() -> Unit = {}): Anchor {
        return addChild(Anchor(href).apply(init))
    }

    fun text(text: String): Text {
        return addChild(Text(text))
    }

    operator fun String.unaryPlus(): Text {
        return this@Node.addChild(Text(this))
    }

    private fun removeChild(childNode: Node) {
        children.remove(childNode)
    }

    fun removeSelf() {
        val parentNode = parent
        if (parentNode != null) {
            parentNode.removeChild(this)
        } else throw Exception("Node doesn't hava a parent")
    }

    fun addAfter(node: Node): Node {
        val oldParent = node.parent
        val parent = this.parent
            ?: throw Exception("Can't add node after another if it has no parent")
        val index = parent.children.indexOf(this)
        if (index == -1) {
            throw Exception("Didn't find object to add after")
        }
        parent.children.add(index + 1, node)
        node.parent = parent
        oldParent?.removeChild(node)
        return node
    }

    fun addBefore(node: Node): Node {
        val oldParent = node.parent
        val parent = this.parent
            ?: throw Exception("Can't add node before another if it has no parent")
        val index = parent.children.indexOf(this)
        if (index == -1) {
            throw Exception("Didn't find object to add after")
        }
        parent.children.add(index, node)
        node.parent = parent
        oldParent?.removeChild(node)
        return node
    }


    // todo abstract ?
    abstract fun write(bw: BackendWriter)

    private fun ArrayList<Node>.fillWithDescendants(node: Node, filter: (Node) -> Boolean) {
        node.children.forEach {
            if (filter(it)) {
                add(it)
            }
            fillWithDescendants(it, filter)
        }
    }

    fun descendant(filter: (Node) -> Boolean = { true }): ArrayList<Node> {
        return arrayListOf<Node>().apply { fillWithDescendants(this@Node, filter) }
    }

    fun children(filter: (Node) -> Boolean = { true }): ArrayList<Node> {
        return arrayListOf<Node>()
            .apply { children.forEach { if (filter(it)) this.add(it) } }
    }

    fun parent(): Node? {
        return parent
    }

    inline fun <reified T : Node> T.roles(vararg role: String): T {
        role.forEach { roles.add(it) }
        return this
    }

    fun toYamlString(): String {
        return ObjectMapper(YAMLFactory().disable(YAMLGenerator.Feature.WRITE_DOC_START_MARKER))
            .writeValueAsString(this)
    }

}

