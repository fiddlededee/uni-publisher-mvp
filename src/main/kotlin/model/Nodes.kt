package model

class DummyNode() : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

abstract class NoWriterNode() : Node() {
    override fun write(bw: BackendWriter) {
        throw Exception ("Writer in NoWriterNodes should be defined in CustomStyle")
    }
}

class Document() : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }

}

class OpenBlock() : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class Col (var width : Width)

enum class WidthUnit { rel, mm }

class Width(var value : Float, var unit : WidthUnit = WidthUnit.rel )

class Table() : Node() {
    val cols = arrayListOf<Col>();
    fun col(width : Width) {
        cols.add(Col(width))
    }
    // TODO: implement cols property
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

enum class TRG {head, body, foot}

class TableRowGroup(val type : TRG) : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class TableRow() : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class TableCell() : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class OrderedList() : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class UnorderedList() : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class ListItem() : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }

    fun wrapListItemContents(): ListItem {
        // Wrapping each inline nodes chains into paragraph node
        fun wrap(itemsToWrap: ArrayList<ArrayList<Node>>) {
            itemsToWrap.forEach { itemsChain ->
                if (itemsChain.isEmpty()) return@forEach
                itemsChain[0].addBefore(Paragraph())
                    .apply { itemsChain.forEach { itemToWrap -> addChild(itemToWrap) } }
            }
        }

        // Finding inline nodes chains
        val itemsToWrap = arrayListOf<ArrayList<Node>>()
        itemsToWrap.add(arrayListOf())
        this.children.forEach { listItemChild ->
            if (listItemChild.isInline) itemsToWrap.last().add(listItemChild)
            else if (itemsToWrap.last().isNotEmpty()) itemsToWrap.add(arrayListOf())
        }
        wrap(itemsToWrap)
        return this
    }
}

class Header(var level: Int) : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class Paragraph() : Node() {
    override val isInline: Boolean get() = false
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class Span() : Node() {
    override val isInline: Boolean get() = true
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class Anchor(var href: String) : Node() {
    override val isInline: Boolean get() = true
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}

class Text(var text: String) : Node() {
    override val isInline: Boolean get() = true
    override fun write(bw: BackendWriter) {
        bw.write(this)
    }
}
