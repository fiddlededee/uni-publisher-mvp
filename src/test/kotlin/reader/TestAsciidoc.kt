package reader

import model.Document
import model.Node
import model.Text
import org.approvaltests.Approvals
import org.asciidoctor.Asciidoctor
import org.asciidoctor.Options
import org.junit.jupiter.api.Test

class TestAsciidoc {
    @Test
    fun testSimpleAsciidoc() {
        """
            = Sample Asciidoc 
           
            == Heading 1
            
            [.unnecessary]
            This paragraph should not appear in result
            
            This paragraph should appear in result
            
            == Heading 2
            
            Some more paragraph
        """
            .trimIndent()
            .asciidocAsHtml()
            .apply {println(this)}
            .htmlMarkupAsHtmlNode()
            .selectAtXpath("/html/body")!!
            .parseWithBasic().apply {
                this.descendant { it.roles.contains("unnecessary") }
                    .forEach { it.removeSelf() }
            }
            .output()
            .apply { Approvals.verify(this) }
    }
}

object AsciidocHtmlFactory {
    private val factory: Asciidoctor = Asciidoctor.Factory.create()
    fun getHtml(string: String): String {
        return factory.convert(
            string, Options.builder().backend("html5")
                .sourcemap(true).build()
        )
    }
}

fun String.asciidocAsHtml(): String {
    return AsciidocHtmlFactory.getHtml(this)
}

fun HtmlNode.parseWithBasic(): Document {
    val document = Document()
    GenericHtmlReader(document, this, setOf("unnecessary")).apply { iterateAll() }
    return document
}


fun String.htmlMarkupAsHtmlNode(): HtmlNode {
    return HtmlNode(this)
}


fun Document.output(): String {
    val astAsText = StringBuilder()
    fun outputNode(node: Node, level: Int): String {
        if (node is Text) {
            astAsText.appendLine("${" ".repeat(level * 2)}text: [${node.text}]")
        } else {
            astAsText.appendLine(
                "${" ".repeat(level * 2)}${node::class.java.simpleName} id=${node.id}" +
                        " roles=${node.roles.joinToString(",")}"
            )
            node.children.forEach {
                outputNode(it, level + 1)
            }
        }
        return astAsText.toString()
    }
    outputNode(this, 0)
    return astAsText.toString()
}
