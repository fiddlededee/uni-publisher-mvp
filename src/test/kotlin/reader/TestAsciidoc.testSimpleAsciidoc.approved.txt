Document id=null roles=
  Header id=_heading_1 roles=
    text: [Heading 1]
  Paragraph id=null roles=
    text: [This paragraph should appear in result]
  Header id=_heading_2 roles=
    text: [Heading 2]
  Paragraph id=null roles=
    text: [Some more paragraph]
